docs(react-native): Upgrade guide (#957)
bugsnag/bugsnag-js: packages/react-native/UPGRADING.md

For iOS, open up the project in XCode `open ios/<ProjectName>.xcworkspace`, and from the menubar select `Product > Clean build folder`.

For Android, run `cd android && ./gradlew clean`.

### Adding `@bugsnag/react-native`

Full instructions are available at the [React Native integration guide](https://docs.bugsnag.com/platforms/react-native/react-native#installation) but here is a summary.

Install the new package from npm using your preferred command:

```sh
npm install --save @bugsnag/react-native
# or
yarn add @bugsnag/react-native
```

#### Android

Add the following to your `android/app/build.gradle` to integrate Bugsnag into your Anroid project:

```groovy
apply from: "../../node_modules/@bugsnag/react-native/bugsnag-react-native.gradle"
```

#### iOS

Run cocoapods again so that it picks up the new local dependency:

```
cd ios
pod install
```

### Updating usage

Full integration instructions are available on the [React Native integration guide](https://docs.bugsnag.com/platforms/react-native/react-native#installation). This section will focus on specific areas that have changed.

> Since this update involved a major bump to Bugsnag's JS, Android and iOS notifiers – bringing all of their APIs in sync, virtually every usage of Bugsnag's API in your application will need updating.

Reviewer: Is there a more positive way to word this?

Author: Have attempted to be more cheery about this. Let me know what you think.

```diff
- #import <BugsnagReactNative/BugsnagReactNative.h>
- #import "BugsnagConfiguration.h"
+ import <Bugsnag/Bugsnag.h>
```

Inside your `application:didFinishLaunchingWithOptions` method, replace any Bugsnag initialization code with the updated version:

```diff
- [BugsnagReactNative start];
+ [Bugsnag start];
```

Or if you need to customize your configuration:

```diff
- BugsnagConfiguration *config = [BugsnagConfiguration new];
- config.releaseStage = @"beta";
- [BugsnagReactNative startWithConfiguration:config];
+ BugsnagConfiguration *config = [BugsnagConfiguration loadConfig];
+ config.releaseStage = @"beta";
+ [Bugsnag startWithConfiguration:config];
```

See the new [Bugsnag Cocoa configuration guide](https://docs.bugsnag.com/platforms/ios/configuration-options/) for all available configuration options.

If you have set your API key in `Info.plist`, the format has changed from a top level string to a dictionary.

```
- <key>BugsnagAPIKey</key>
- <string>YOUR_API_KEY</string>
+ <key>bugsnag</key>
+ <dict>
+     <key>apiKey</key>
+     <string>YOUR_API_KEY</string>
+ </dict>

> ```

Reviewer: We want to emphasise use of the plist/manifest as much as possible - can you move this to after the `[Bugsnag start]` snippet and reference the new ability to set other configuration options? Then have the in-code snippet afterwards.

##### Android

Replace the following import which will likely be in your `MainApplication.java` file:

```diff
- import com.bugsnag.android.BugsnagReactNative;
+ import com.bugsnag.android.Bugsnag;
```

> Inside your application's `onCreate` method, replace any Bugsnag initialization code with the updated version:

Reviewer: As above, it would be good to reference the file-loading config mechanism here.

```diff
- Configuration config = new Configuration();
- config.setReleaseStage("beta");
- BugsnagReactNative.startWithConfiguration(this /* app context */, config);
+ Configuration config = Configuration.load(this);
+ config.setReleaseStage("beta");
Bugsnag.start(this /* app context */, config);
```

###### JS

> The JS interface has been updated to match our new notifier spec standard – that means it looks the same as our JS notifier, and uses the same names and concepts as our other platforms.

Reviewer: Do we want to refer to the spec?

```suggestion
As part of the `bugsnag-js monorepo`, the JS interface now looks the same as our JS notifier, and uses the same names and concepts as our other platforms.
```

Author: I figured it was ok to refer to in passing, but happy with your suggestion.

> When the JS layer loads, __the native layers and notifiers must already have been configured__. It you attempt to initialize the JS layer without having first configured the native layer, it __will not work__. This is new behaviour – previously we allowed configuration of the native layer to be done lazily and automatically, but this meant native errors on startup could be missed.

Reviewer: ```suggestion
When the JS layer loads, __the native iOS/Android Bugsnag client must already have been configured__. If you attempt to initialize the JS layer without having first configured the native layer, it __will not work__. This is new behavior – previously we allowed configuration of the native layer to be done lazily and automatically, but this meant native errors on startup could be missed.
```

> Since the JS layer hooks in to an already running native layer, there are fewer configuration options available. Most configuration must be done in the native layer, with only a small subset of JS-specific options allowed in the JS layer.

Reviewer: Move underneath proceeding snippet as it's referring to configuration, not starting?

```diff
- import { Client } from 'bugsnag-react-native';
- const client = new Client();
+ import Bugsnag from '@bugsnag/react-native'
+ Bugsnag.start()
```

JS configuration options are now supplied directly to the `start` method:

```diff
- import { Configuration } from 'bugsnag-react-native'
- const config = new Configuration()
- config.codeBundleId = '1.2.3-r091283'
- const client = new Client(config)
+ Bugsnag.start({
+   codeBundleId: '1.2.3-r091283'
+ })
```

See the [React Native configuration options page](https//docs.bugsnag.com/platforms/react-native/react-native/configuration-options) for reference.

#### General usage

> ##### iOS

Reviewer: The subheadings are getting quite nested here, to the extent that the bold body text stands out more than the heading.

Could we have an H5 referring to native usage followed by H5's for each of the JS sections?

Author: I think I have addressed this.

##### Android

Bugsnag React Native vendors the `bugsnag-android` library, so the usage is the same whether you are using `bugsnag-android` directly, or whether you are using Bugsnag in the Android side of your React Native project.

See the [V4 to V5 section of the `bugsnag-android` upgrade guide](https://github.com/bugsnag/bugsnag-android/blob/master/UPGRADING.md#upgrade-from-4x-to-5x) for examples of how to update the usage of Bugsnag's APIs.

##### JS

Previously initialization of Bugsnag would return a `Client` that you would need to hold a reference to and pass around your application. You can still do that if you want, but importing `Bugsnag` is a static namespace that you can import and use anywhere, provided `Bugsnag.start()` has been called somewhere first.

```diff
- const client = new Client()
+ Bugsnag.start()
```

###### Handled errors

To report a handled error call the `notify()` method:

```diff
+ client.notify(new Error('uh oh'))
+ Bugsnag.notify(new Error('uh oh'))
```

> The second argument is an `onError` callback which will receive the error report that will be sent to Bugsnag. This is similar, but previously the data structure was a `report` – now it is an `event`. See the [Customizing error reports](https://docs.bugsnag.com/platforms/react-native/react-native/customizing-error-reports/#updating-events-using-callbacks) for full information. An example of  adding metadata is shown below:

Reviewer: ```suggestion
The second argument is an `onError` callback which will receive the error report that will be sent to Bugsnag. This is similar to the previous syntax, but previously the data structure was a `report` – now it is an `event`. See the [Customizing error reports](https://docs.bugsnag.com/platforms/react-native/react-native/customizing-error-reports/#updating-events-using-callbacks) for full information. An example of adding metadata to an event is shown below:
```

```diff
- client.notify(err, (report, originalError) => {
+ Bugsnag.notify(err, event => {
-   report.metadata = {
-     account: { id: '123', name: 'Acme Co' }
-   }
+   event.addMetadata('account', { id: '123', name: 'Acme Co' })

    // Note, originalError error now exists at event.originalError
  })
```

###### Adding/removing callbacks

```diff
- configuration.registerBeforeSendCallback(fn)
- configuration.unregisterBeforeSendCallback(fn)
+ Bugsnag.start({ onError: fn })
  // or
+ Bugsnag.addOnError(fn)
+ Bugsnag.removeOnError(fn)
```

> 

Reviewer: Add a quick sentence to go with the snippet and link to docs?

###### Adding user information

```diff
- client.setUser('1234', 'Jessica Jones', 'jess@example.com')
+ Bugsnag.setUser('1234', 'jess@example.com', 'Jessica Jones')
```

__Note: the argument order has changed__ from `id,name,email` to `id,email,name`

> 

Reviewer: Add a quick sentence to go with the snippet and link to docs?

###### Sessions

```diff
- client.startSession()
- client.pauseSession()
- client.resumeSession()
+ Bugsnag.startSession()
+ Bugsnag.pauseSession()
+ Bugsnag.resumeSession()
```

> 

Reviewer: Add a quick sentence to go with the snippet and link to docs?
